src/webapi/event.rs

```diff
@@ -1221,6 +1223,290 @@ impl ConcreteEvent for PopStateEvent {
+    /// navigating away from the page that opened the connection.
+    pub const GOING_AWAY: CloseEventCode = CloseEventCode(1001);
+    /// The endpoint is terminating the connection due to a protocol error.
+    pub const PROTOCOL_ERROR: CloseEventCode = CloseEventCode(1002);
+    /// The connection is being terminated because the endpoint received data of a type it cannot
+    /// accept (for example, a text-only endpoint received binary data).
+    pub const UNSUPPORTED_DATA: CloseEventCode = CloseEventCode(1003);
+    /// Reserved. Indicates that no status code was provided even though one was expected.
+    pub const NO_STATUS_RECEIVED: CloseEventCode = CloseEventCode(1005);
+    /// Reserved. Used to indicate that a connection was closed abnormally (that is, with no close
+    /// frame being sent) when a status code is expected.
+    pub const ABNORMAL_CLOSURE: CloseEventCode = CloseEventCode(1006);
+    /// The endpoint is terminating the connection because a message was received that contained
+    /// inconsistent data (e.g., non-UTF-8 data within a text message).
+    pub const INVALID_FRAME_PAYLOAD_DATA: CloseEventCode = CloseEventCode(1007);
+    /// The endpoint is terminating the connection because it received a message that violates its
+    /// policy. This is a generic status code, used when codes 1003 and 1009 are not suitable.
+    pub const POLICY_VIOLATION: CloseEventCode = CloseEventCode(1008);
+    /// The endpoint is terminating the connection because a data frame was received that is too
+    /// large.
+    pub const MESSAGE_TOO_BIG: CloseEventCode = CloseEventCode(1009);
+    /// The client is terminating the connection because it expected the server to negotiate one or
+    /// more extensions, but the server didn't.
+    pub const MISSING_EXTENSION: CloseEventCode = CloseEventCode(1010);
+    /// The server is terminating the connection because it encountered an unexpected condition
+    /// that prevented it from fulfilling the request.
+    pub const INTERNAL_ERROR: CloseEventCode = CloseEventCode(1011);
+    /// The server is terminating the connection because it is restarting.
+    pub const SERVICE_RESTART: CloseEventCode = CloseEventCode(1012);
+    /// The server is terminating the connection due to a temporary condition, e.g. it is
+    /// overloaded and is casting off some of its clients.
+    pub const TRY_AGAIN_LATER: CloseEventCode = CloseEventCode(1013);
+    /// The server was acting as a gateway or proxy and received an invalid response from the
+    /// upstream server. This is similar to 502 HTTP Status Code.
+    pub const BAD_GATEWAY: CloseEventCode = CloseEventCode(1014);
+    /// Reserved. Indicates that the connection was closed due to a failure to perform a TLS
+    /// handshake (e.g., the server certificate can't be verified).
+    pub const TLS_HANDSHAKE: CloseEventCode = CloseEventCode(1015);
+}
```

Comment:
Few things:
1) Matching works fine with the struct as it is defined now
2) Other codes are allowed to be used (3000-4999 are either library or application defined)
3) Using an enum uses more than 16 bits of space required to store the code
4) Using an enum means you have to worry about the "__Other" variant taking a value defined as a named code

I think overall using a public newtype maps much more nicely onto the set of values.

Reply:
> Matching works fine with the struct as it is defined now

One thing I really don't like when matching constants defined this way is something like this:

Let's say you have a `use` for the constants you want to match on, but then you temporairly comment it out and forget about it. Something like this:

```rust
// use another::workspace::{CONSTANT_A, CONSTANT_B};
// ^^^^ - oops, I commented this out and forgot about it while refactoring

match value {
    CONSTANT_A => { /* do something */ },
    CONSTANT_B => { /* do something */ },
    _ => /* do something else */
}
```
Rust *will* warn you that you have multiple wildcard matches, but if you're like me and you only clean up warnings **after** refactoring your code (since usually any significant refactoring results in a **lot** of warnings about unused imports and such) then you might not notice that you're in fact doing multiple wildcard matches and scratch your head why it doesn't work.

With enums you usually match like this:

```rust
match value {
    EnumKind::ConstantA => {},
    EnumKind::ConstantB => {},
    _ => {}
}
````

which doesn't have the same problem.

> Other codes are allowed to be used (3000-4999 are either library or application defined)

Yep. I know. Both approaches allow for them. It's just that with the `enum` approach the custom codes are encapsulated into an opaque type.

> Using an enum uses more than 16 bits of space required to store the code

This isn't really an issue to be quite honest; we're not storing it anywhere, and even if we were it's not a big deal.

> Using an enum means you have to worry about the "__Other" variant taking a value defined as a named code

You don't. That's the point of making that variant `#[doc(hidden)]`. Either the user uses one of the named variants, **or** the user has to convert it to `u16` and handle it raw. With this approach for those users who don't care about custom codes it's possible to easily match on all known standard status codes with a `match`, and you get a `Debug` implementation for free.

Perhaps someone else could chime in which version they prefer? I don't know, maybe @CryZe or @Pauan - which version do you prefer?

Reply:
> Let's say you have a use for the constants you want to match on, but then you temporairly comment it out and forget about it. Something like this:

There's no difference here? Both enums and structs with associated consts behave identically in this situation. If you import the variants/associated consts directly, then you suffer the issue you described, if you import the type, then you don't.

> You don't. That's the point of making that variant #[doc(hidden)]

In your example, you implemented `From<u16>` - that means you have to check whether the value being converted is one of the "known" constants, and convert it specially if so. If you were intending to just always convert it to the `__Other` variant, then I think having two way to represent a single code is a really bad idea.

Reply:
> There's no difference here? Both enums and structs with associated consts behave identically in this situation. If you import the variants/associated consts directly, then you suffer the issue you described, if you import the type, then you don't.

Technically, yes. But in practice I generally tend to see consts imported directly (e.g. `use foo::bar::CONSTANT`), while for enums people usually only import the type (e.g. `use foo::bar::Enum`).

> In your example, you implemented `From<u16>` - that means you have to check whether the value being converted is one of the "known" constants, and convert it specially if so. If you were intending to just always convert it to the __Other variant, then I think having two way to represent a single code is a really bad idea.

The point is that from the user's perspective there would be only one way to represent a single code, since the user doesn't know about the `__Other` variant! (Or in other words - the `__Other` variant is our private implementation detail.) This is almost exactly the same as [non-exhaustive enums](https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md), except the non-exhaustive case has a value attached to it which the user can extract to meaningfully handle codes which `stdweb` doesn't know about in a way that is both backwards and forwards compatible.

```rust
match code {
    // For codes we know about the user can handle them in a Rust-y way:
    SocketClosedCode::NormalClosure => { ... },
    SocketClosedCode::GoingAway => { ... },
    // For codes we don't know the user can handle them theirselves:
    code if code.as_raw() == 4000 => { ... },
    // The user *must* have a wildcard match (since `__Other` is hidden), so we can add new codes without breaking anyone's code.
    code => { println!( "Unhandled result: {:?}", code ); }
}
```

To me this feels a lot more Rust-y compared to using raw constants, no?

Reply:
> Technically, yes. But in practice I generally tend to see consts imported directly

For global constants, maybe? For associated constants, no.

> The point is that from the user's perspective there would be only one way to represent a single code, since the user doesn't know about the __Other variant!

I feel like we're talking past each other? The user has to be able to construct any u16 code, which means we have to implement that conversion: when we implement that conversion, we'd have to check whether the code is one of the "known" codes, and map it to the corresponding enum variant instead of `__Other`. If we don't do that, then there are multiple ways to represent a single code!

This also means that users can define their own codes:

```
pub const MyCloseCode: CloseEventCode = CloseEventCode(4001);

match code {
    CloseEventCode::NormalClosure => { ... },
    MyCloseCode => { ... },
    _ => { ... }
}
```

Reply:
> For global constants, maybe? For associated constants, no.

Ah, yes, you're actually right here.

> when we implement that conversion, we'd have to check whether the code is one of the "known" codes, and map it to the corresponding enum variant instead of `__Other`

Yes, of course!

> This also means that users can define their own codes

This is also true if we keep it as an `enum`.

```
// Once `const fn`s are stabilized:
pub const MyCloseCode: SocketClosedCode = SocketClosedCode::new(4001);
// Before `const fn`:
pub const MyCloseCode: u16 = 4001;
```

Although yes, they wouldn't be able to extend the enum itself.

Anyhow, although I feel the `enum` is somewhat nicer in general I'm not really strongly convinced it's the better option. To me both of them are kind-of equivalent and differ mostly stylistically.

If we'd go with the constant variant I'd like to see a `Debug` implementation which would print out the codes in a textual format.

Reply:
Also, perhaps it would make sense to make the conversion from `u16` to the status code type fallible? (E.g. like the `http` crate does with HTTP status code.)

Then we wouldn't have to handle the exception for invalid status code in `close()`.

Reply:
I don't feel particularly strongly about that, but it would be a shame to lose the ability to construct constant codes given that `const fn` is still unstable. Maybe something to reconsider before 1.0 since it may be stabilised by then, but I think for now it just makes it more painful to use for little benefit - how many bugs are we really going to prevent by stopping people from constructing bogus close codes?

Reply:
I will definitely add a `Debug` implementation though.

Reply:
Yeah, not being able to construct them statically is one of the downsides. If `const fn`s were stable it would be a no-brainer, but as-is I don't really feel strongly either way since it doesn't seem like a major enough of a win to warrant the downsides.